proto/go_proto_library.bzl

```diff
@@ -0,0 +1,238 @@
+
+_DEFAULT_LIB = "go_default_library"  # matching go_library
+
+_PROTOS_SUFFIX = "_protos"
+
+def _go_prefix(ctx):
+  """slash terminated go-prefix."""
+  prefix = ctx.attr.go_prefix.go_prefix
+  if prefix and not prefix.endswith("/"):
+    prefix = prefix + "/"
+  return prefix
+
+def _exernal_dir(files):
+  if not files:
+    return []
+  toks = list(files)[0].dirname.split("/")
+  if not toks or toks[0] != "external":
+    return []
+  return "/".join(toks[:2])
+
+def _proto_import(deps):
+  "Compute any needed -I options to protoc from external filegroups."
+  return [_exernal_dir(d.files) for d in deps if not hasattr(d, "_protos")]
+
+def _go_proto_library_gen_impl(ctx):
+  """Rule implementation that generates Go using protoc."""
+  bazel_style = ctx.label.name != _DEFAULT_LIB + _PROTOS_SUFFIX
+  protos = list(ctx.files.src)
+  go_package_name = ""
+  if bazel_style:
+    go_package_name = "/" + ctx.label.name[:-len(_PROTOS_SUFFIX)]
+  m_import_path = ",".join(["M%s=%s%s%s" % (f.path, _go_prefix(ctx),
+                                            ctx.label.package, go_package_name)
+                            for f in ctx.files.src])
+  for d in ctx.attr.deps:
+    if not hasattr(d, "_protos"):
+      # should be a raw filegroup then
+      protos += list(d.files)
+      continue
+    protos += d._protos
+    m_import_path += "," + d._m_import_path
+  use_grpc = ""
+  if ctx.attr.grpc:
+    use_grpc = "plugins=grpc,"
+
+  offset = 0
+  proto_out = ctx.outputs.out
+  if bazel_style:
+    offset = -1  # extra directory added, need to remove
+    proto_out = ctx.new_file(
+        ctx.configuration.bin_dir, ctx.files.src[0].basename[:-5] + "pb.go")
+  outdir = "/".join(
+      ctx.outputs.out.dirname.split("/")[:offset-len(ctx.label.package.split("/"))])
+  ctx.action(
+      inputs=protos + ctx.files.protoc_gen_go,
+      outputs=[proto_out],
+      arguments=["-I.", "--go_out=%s%s:%s" % (
+          use_grpc, m_import_path, outdir)] + [
+              "-I"+i for i in _proto_import(ctx.attr.deps)
+          ] + [
+              f.path for f in ctx.files.src
+          ],
+      progress_message="Generating into %s" % ctx.outputs.out.short_path,
+      mnemonic="GoProtocGen",
+      env = {"PATH": ctx.files.protoc_gen_go[0].dirname},
+      executable=ctx.executable.protoc)
+  # This is the current hack for files without 'option go_package'
+  # Generate into .pb.go, then cp into "real location"
+  if bazel_style:
+    ctx.action(
+        inputs=[proto_out],
+      outputs=[ctx.outputs.out],
+        command="cp %s %s" % (proto_out.path, ctx.outputs.out.path),
+        mnemonic="GoProtocGenCp")
+  return struct(_protos=protos,
+                _m_import_path=m_import_path,
+                name=(ctx.label.package + go_package_name))
+
+go_proto_library_gen = rule(
+    attrs = {
+        "deps": attr.label_list(),
+        "src": attr.label(
+            mandatory = True,
+            single_file = True,
```

Comment:
a srcs list matches the style of other *_proto_library -- for now I only accept 1 src, but I can imagine changing/fixing that.
The reason is one only of current needs.

As to the other thing, yes you need all transitive protos in the action at generation time and their import mapping, but that is handled via the '_protos' rule provider.